src/coreclr/runtime/amd64/AllocFast.asm

```diff
@@ -20,16 +20,16 @@ LEAF_ENTRY RhpNewFast, _TEXT
         ;;
         ;; eax: base size
         ;; rcx: MethodTable pointer
-        ;; rdx: Thread pointer
+        ;; rdx: ee_alloc_context pointer
         ;;
 
-        mov         rax, [rdx + OFFSETOF__Thread__m_alloc_context__alloc_ptr]
+        mov         rax, [rdx + OFFSETOF__ee_alloc_context__alloc_ptr]
         add         r8, rax
-        cmp         r8, [rdx + OFFSETOF__Thread__m_eeAllocContext__combined_limit]
```

Comment:
Similar to 32-bit case, this assumes that there won't ever be an OSes that gives end of address space to user mode. Is it a safe assumption to make?

Reply:
None of the currently supported 64-bit platforms actually has full 64-bit address space. The top bits are prevalently sign extended and/or used as pointer authentication.

On Windows (x64 and arm64) and Linux (arm64) the address space is divided in such a way that we should not run into overflow issues.

I didn't check Linux x64, rv64 and la64 but presumably they do the user address space split in such a way that this is not an issue.

macOS also has a limit on user virtual address space that makes the overflow impossible.

(Historically only x86 Linux and macOS had a mode where top half of the address space was available to user mode apps, and I mean nearly the full 4Gb. Windows x86 had the 3Gb/1Gb split. These are odd exceptions though.)

Reply:
If you feel like it's worth converting everything to the pattern used in https://github.com/dotnet/runtime/pull/115339#discussion_r2118295714, I can give it a try. It will just need a lot of retesting.

Reply:
LA64: User mode addresses end at `0x3FFF_FFFF_FFFF_FFFF`, followed by a hole and then kernel address space starting at `0xC000_0000_0000_0000`
RV64: Depends on Linux kernel configuration (https://docs.kernel.org/6.5/riscv/vm-layout.html). Even at the largest one (SV57) the user mode addresses end at `0x00FF_FFFF_FFFF_FFFF`.
Apple/arm64: Maximum address of allocation is `0x0000_7FFF_FE00_0000`.
Linux/arm64: Depends on Linux kernel configuration (https://www.kernel.org/doc/html/v5.8/arm64/memory.html). With 64Kb pages the largest user mode address is `0x000F_FFFF_FFFF_FFFF`

Reply:
Done for x64 and amd64 so far.

> I understand that there may not be an actual problem on currently supported OSes, but that's a weak argument when it comes to security defense in depth.

That's fair and I agree. Anything that's easier to reason about is preferable.